crates/matrix-sdk/src/event_cache/linked_chunk.rs

```diff
@@ -43,48 +38,156 @@ pub enum LinkedChunkError {
+/// The checks are per-field, not per-struct. That's why `first` and `last` have
+/// been extracted in a separate type, this type.
+struct LinkedChunkLinks<const CHUNK_CAPACITY: usize, Item, Gap> {
+    /// The first chunk.
+    first: NonNull<Chunk<CHUNK_CAPACITY, Item, Gap>>,
+    /// The last chunk.
+    last: Option<NonNull<Chunk<CHUNK_CAPACITY, Item, Gap>>>,
+}
+
+impl<const CAP: usize, Item, Gap> LinkedChunkLinks<CAP, Item, Gap> {
+    /// Get the first chunk, as an immutable reference.
+    fn first_chunk(&self) -> &Chunk<CAP, Item, Gap> {
+        unsafe { self.first.as_ref() }
+    }
+
+    /// Get the latest chunk, as an immutable reference.
+    fn latest_chunk(&self) -> &Chunk<CAP, Item, Gap> {
+        unsafe { self.last.unwrap_or(self.first).as_ref() }
+    }
+
+    /// Get the latest chunk, as a mutable reference.
+    fn latest_chunk_mut(&mut self) -> &mut Chunk<CAP, Item, Gap> {
+        unsafe { self.last.as_mut().unwrap_or(&mut self.first).as_mut() }
+    }
+
+    /// Get the chunk as a reference, from its identifier, if it exists.
+    fn chunk(&self, identifier: ChunkIdentifier) -> Option<&Chunk<CAP, Item, Gap>> {
+        let mut chunk = self.latest_chunk();
+
+        loop {
+            if chunk.identifier() == identifier {
+                return Some(chunk);
+            }
+
+            chunk = chunk.previous()?;
+        }
+    }
+
+    /// Get the chunk as a mutable reference, from its identifier, if it exists.
+    fn chunk_mut(&mut self, identifier: ChunkIdentifier) -> Option<&mut Chunk<CAP, Item, Gap>> {
+        let mut chunk = self.latest_chunk_mut();
+
+        loop {
+            if chunk.identifier() == identifier {
+                return Some(chunk);
+            }
+
+            chunk = chunk.previous_mut()?;
+        }
+    }
+}
+
 /// The [`LinkedChunk`] structure.
 ///
 /// It is similar to a linked list, except that it contains many items `Item`
 /// instead of a single one. A chunk has a maximum capacity of `CHUNK_CAPACITY`.
 /// Once a chunk is full, a new chunk is created. Not all chunks are necessarily
 /// entirely full. A chunk can represents a `Gap` between other chunks.
-pub struct LinkedChunk<const CHUNK_CAPACITY: usize, Item, Gap, Listener = ()> {
-    /// The first chunk.
-    first: NonNull<Chunk<CHUNK_CAPACITY, Item, Gap>>,
-    /// The last chunk.
-    last: Option<NonNull<Chunk<CHUNK_CAPACITY, Item, Gap>>>,
+///
+/// ⚠️ When the `LinkedChunk` is manipulated, all updates (i.e. operations) are
+/// kept in memory. It's like a history of all `LinkedChunk`'s updates. This is
+/// useful when one wants to replicate the `LinkedChunk` in another kind of
+/// storage, like a database. Every time one calls `LinkedChunk::updates()` to
+/// read the updates, it's mandatory to clean, to drain, to empty it. Even if
+/// one doesn't care about the updates, it's necessary to empty the updates,
+/// otherwise it will grow in memory for the eternity.
```

Comment:
Sounds like an API footgun. Could `update_history` be optionally set, and a method like `enable_history()` would assign it to `Some(Vec::new())`; then instead of recording each update, we'd do it only if the `update_history.is_some()`?

Reply:
The `new` constructor disables the update history, whilst the `new_with_update_history` enables it. The `updates` method returns an `Option<&mut Vec<…>>` now.